File: facebook-core/src/main/java/com/facebook/internal/FeatureManager.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.internal;

import android.content.Context;
import androidx.annotation.RestrictTo;

import com.facebook.FacebookSdk;

import java.util.HashMap;
import java.util.Map;

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public final class FeatureManager {

    private static final String FEATURE_MANAGER_STORE = "com.facebook.internal.FEATURE_MANAGER";

    private static final Map<FeatureManager.Feature, String[]> featureMapping = new HashMap<>();

    public static void checkFeature(final Feature feature, final Callback callback) {
        FetchedAppGateKeepersManager.loadAppGateKeepersAsync(new FetchedAppGateKeepersManager.Callback() {
            @Override
            public void onCompleted() {
                callback.onCompleted(FeatureManager.isEnabled(feature));
            }
        });
    }

    public static boolean isEnabled(Feature feature) {
        if (Feature.Unknown == feature) {
            return false;
        }

        if (Feature.Core == feature) {
            return true;
        }

        String version = FacebookSdk.getApplicationContext()
                .getSharedPreferences(FEATURE_MANAGER_STORE, Context.MODE_PRIVATE)
                .getString(feature.toKey(), null);
        if (version != null && version.equals(FacebookSdk.getSdkVersion())) {
            return false;
        }

        Feature parent = feature.getParent();
        if (parent == feature) {
            return getGKStatus(feature);
        } else {
            return isEnabled(parent) && getGKStatus(feature);
        }
    }

    public static void disableFeature(Feature feature) {
        FacebookSdk.getApplicationContext()
                .getSharedPreferences(FEATURE_MANAGER_STORE, Context.MODE_PRIVATE)
                .edit()
                .putString(feature.toKey(), FacebookSdk.getSdkVersion())
                .apply();
    }

    public static FeatureManager.Feature getFeature(String className) {
        initializeFeatureMapping();
        for (final Map.Entry<FeatureManager.Feature, String[]> entry : featureMapping.entrySet()) {
            for (String v : entry.getValue()) {
                if (className.startsWith(v)) {
                    return entry.getKey();
                }
            }
        }
        return FeatureManager.Feature.Unknown;
    }

    private synchronized static void initializeFeatureMapping() {
        if (!featureMapping.isEmpty()) {
            return;
        }

        featureMapping.put(FeatureManager.Feature.AAM, new String[]{
                "com.facebook.appevents.aam.",
        });
        featureMapping.put(FeatureManager.Feature.CodelessEvents, new String[]{
                "com.facebook.appevents.codeless.",
        });
        featureMapping.put(FeatureManager.Feature.ErrorReport, new String[]{
                "com.facebook.internal.instrument.errorreport.",
        });
        featureMapping.put(FeatureManager.Feature.PrivacyProtection, new String[]{
                "com.facebook.appevents.ml.",
        });
        featureMapping.put(FeatureManager.Feature.SuggestedEvents, new String[]{
                "com.facebook.appevents.suggestedevents.",
        });
        featureMapping.put(FeatureManager.Feature.RestrictiveDataFiltering, new String[]{
                "com.facebook.appevents.restrictivedatafilter.RestrictiveDataManager",
        });
        featureMapping.put(FeatureManager.Feature.IntelligentIntegrity, new String[]{
                "com.facebook.appevents.integrity.IntegrityManager",
        });
        featureMapping.put(FeatureManager.Feature.EventDeactivation, new String[]{
                "com.facebook.appevents.eventdeactivation.",
        });
        featureMapping.put(FeatureManager.Feature.Monitoring, new String[]{
                "com.facebook.internal.logging.monitor",
        });
    }

    private static boolean getGKStatus(Feature feature) {
        boolean defaultStatus = defaultStatus(feature);

        return FetchedAppGateKeepersManager.getGateKeeperForKey(
                feature.toKey(),
                FacebookSdk.getApplicationId(),
                defaultStatus);
    }

    private static boolean defaultStatus(Feature feature) {
        switch (feature) {
            case RestrictiveDataFiltering:
            case Instrument:
            case CrashReport:
            case CrashShield:
            case ThreadCheck:
            case ErrorReport:
            case AAM:
            case PrivacyProtection:
            case SuggestedEvents:
            case IntelligentIntegrity:
            case ModelRequest:
            case EventDeactivation:
            case ChromeCustomTabsPrefetching:
            case Monitoring:
            case IgnoreAppSwitchToLoggedOut:
                return false;
            default:
                return true;
        }
    }

    /**
     Feature enum
     Defines features in SDK

     Sample:
     AppEvents = 0x00010000,
                    ^ ^ ^ ^
                    | | | |
                  kit | | |
                feature | |
              sub-feature |
            sub-sub-feature
     1st byte: kit
     2nd byte: feature
     3rd byte: sub-feature
     4th byte: sub-sub-feature
     */
    public enum Feature {
        Unknown(-1),
        // Features in CoreKit
        /** Essential of CoreKit */
        Core(0x00000000),

        AppEvents(0x00010000),
        CodelessEvents(0x00010100),
        RestrictiveDataFiltering(0x00010200),
        AAM(0x00010300),
        PrivacyProtection(0x00010400),
        SuggestedEvents(0x00010401),
        IntelligentIntegrity(0x00010402),
        ModelRequest(0x00010403),
        EventDeactivation(0x00010500),

        Instrument(0x00020000),
        CrashReport(0x00020100),
        CrashShield(0x00020101),
        ThreadCheck(0x00020102),
        ErrorReport(0x00020200),

        Monitoring(0x00030000),

        // Features in LoginKit
        /** Essential of LoginKit */
        Login(0x01000000),
        ChromeCustomTabsPrefetching(0x01010000),
        IgnoreAppSwitchToLoggedOut(0x01020000),

        // Features in ShareKit
        /** Essential of ShareKit */
        Share(0x02000000),

        // Features in PlacesKit
        /** Essential of PlacesKit */
        Places(0x03000000);

        private final int code;

        Feature(int code) {
            this.code = code;
        }

        @Override
        public String toString() {
            String name = "unknown";

            switch (this) {
                case Core: name = "CoreKit"; break;
                case AppEvents: name = "AppEvents"; break;
                case CodelessEvents: name = "CodelessEvents"; break;
                case RestrictiveDataFiltering: name = "RestrictiveDataFiltering"; break;
                case Instrument: name = "Instrument"; break;
                case CrashReport: name = "CrashReport"; break;
                case CrashShield: name = "CrashShield"; break;
                case ThreadCheck: name = "ThreadCheck"; break;
                case ErrorReport: name = "ErrorReport"; break;
                case AAM: name = "AAM"; break;
                case PrivacyProtection: name = "PrivacyProtection"; break;
                case SuggestedEvents: name = "SuggestedEvents"; break;
                case IntelligentIntegrity: name = "IntelligentIntegrity"; break;
                case ModelRequest: name = "ModelRequest"; break;
                case EventDeactivation: name = "EventDeactivation"; break;
                case Monitoring: name = "Monitoring"; break;

                case Login: name = "LoginKit"; break;
                case ChromeCustomTabsPrefetching: name = "ChromeCustomTabsPrefetching"; break;
                case IgnoreAppSwitchToLoggedOut: name = "IgnoreAppSwitchToLoggedOut"; break;

                case Share: name = "ShareKit"; break;

                case Places: name = "PlacesKit"; break;
            }

            return name;
        }

        static Feature fromInt(int code) {
            for (Feature feature : Feature.values()) {
                if (feature.code == code) {
                    return feature;
                }
            }

            return Feature.Unknown;
        }

        String toKey() {
            return "FBSDKFeature" + this.toString();
        }

        public Feature getParent() {
            if ((this.code & 0xFF) > 0) {
                return fromInt(this.code & 0xFFFFFF00);
            } else if ((this.code & 0xFF00) > 0) {
                return fromInt(this.code & 0xFFFF0000);
            } else if ((this.code & 0xFF0000) > 0) {
                return fromInt(this.code & 0xFF000000);
            } else {
                return fromInt(0);
            }
        }
    }

    /**
     * Callback for fetching feature status. Method
     * {@link FeatureManager#checkFeature(Feature, Callback)}} will call GateKeeper manager to load
     * the latest GKs first and then run the callback function.
     */
    public interface Callback {
        /**
         * The method that will be called when the feature status request completes.
         */
        void onCompleted(boolean enabled);
    }
}

File: facebook-core/src/main/java/com/facebook/internal/FeatureManager.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal

import android.content.Context
import androidx.annotation.RestrictTo
import com.facebook.FacebookSdk

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
object FeatureManager {
    private const val FEATURE_MANAGER_STORE = "com.facebook.internal.FEATURE_MANAGER"
    private val featureMapping: MutableMap<Feature, Array<String>> = hashMapOf()

    @JvmStatic
    fun checkFeature(feature: Feature, callback: Callback) {
        FetchedAppGateKeepersManager.loadAppGateKeepersAsync(
            object : FetchedAppGateKeepersManager.Callback {
                override fun onCompleted() {
                    callback.onCompleted(isEnabled(feature))
                }
            })
    }

    @JvmStatic
    fun isEnabled(feature: Feature): Boolean {
        if (Feature.Unknown == feature) {
            return false
        }
        if (Feature.Core == feature) {
            return true
        }
        val version =
            FacebookSdk.getApplicationContext()
                .getSharedPreferences(FEATURE_MANAGER_STORE, Context.MODE_PRIVATE)
                .getString(feature.toKey(), null)
        if (version != null && version == FacebookSdk.getSdkVersion()) {
            return false
        }
        val parent = feature.parent
        return if (parent == feature) {
            getGKStatus(feature)
        } else {
            isEnabled(parent) && getGKStatus(feature)
        }
    }

    @JvmStatic
    fun disableFeature(feature: Feature) {
        FacebookSdk.getApplicationContext()
            .getSharedPreferences(FEATURE_MANAGER_STORE, Context.MODE_PRIVATE)
            .edit()
            .putString(feature.toKey(), FacebookSdk.getSdkVersion())
            .apply()
    }

    @JvmStatic
    fun getFeature(className: String): Feature {
        initializeFeatureMapping()
        for ((key, value) in featureMapping) {
            for (v in value) {
                if (className.startsWith(v)) {
                    return key
                }
            }
        }
        return Feature.Unknown
    }

    @Synchronized
    private fun initializeFeatureMapping() {
        if (!featureMapping.isEmpty()) {
            return
        }
        featureMapping[Feature.AAM] = arrayOf("com.facebook.appevents.aam.")
        featureMapping[Feature.CodelessEvents] = arrayOf("com.facebook.appevents.codeless.")
        featureMapping[Feature.CloudBridge] = arrayOf("com.facebook.appevents.cloudbridge.")
        featureMapping[Feature.ErrorReport] =
            arrayOf("com.facebook.internal.instrument.errorreport.")
        featureMapping[Feature.AnrReport] = arrayOf("com.facebook.internal.instrument.anrreport.")
        featureMapping[Feature.PrivacyProtection] = arrayOf("com.facebook.appevents.ml.")
        featureMapping[Feature.SuggestedEvents] = arrayOf("com.facebook.appevents.suggestedevents.")
        featureMapping[Feature.RestrictiveDataFiltering] =
            arrayOf("com.facebook.appevents.restrictivedatafilter.RestrictiveDataManager")
        featureMapping[Feature.IntelligentIntegrity] =
            arrayOf("com.facebook.appevents.integrity.IntegrityManager")
        featureMapping[Feature.ProtectedMode] =
            arrayOf("com.facebook.appevents.integrity.ProtectedModeManager")
        featureMapping[Feature.MACARuleMatching] =
            arrayOf("com.facebook.appevents.integrity.MACARuleMatchingManager")
        featureMapping[Feature.BlocklistEvents] =
            arrayOf("com.facebook.appevents.integrity.BlocklistEventsManager")
        featureMapping[Feature.FilterRedactedEvents] =
            arrayOf("com.facebook.appevents.integrity.RedactedEventsManager")
        featureMapping[Feature.FilterSensitiveParams] =
            arrayOf("com.facebook.appevents.integrity.SensitiveParamsManager")
        featureMapping[Feature.EventDeactivation] =
            arrayOf("com.facebook.appevents.eventdeactivation.")
        featureMapping[Feature.OnDeviceEventProcessing] =
            arrayOf("com.facebook.appevents.ondeviceprocessing.")
        featureMapping[Feature.IapLogging] = arrayOf("com.facebook.appevents.iap.")
        featureMapping[Feature.Monitoring] = arrayOf("com.facebook.internal.logging.monitor")
    }

    private fun getGKStatus(feature: Feature): Boolean {
        val defaultStatus = defaultStatus(feature)
        return FetchedAppGateKeepersManager.getGateKeeperForKey(
            feature.toKey(), FacebookSdk.getApplicationId(), defaultStatus
        )
    }

    private fun defaultStatus(feature: Feature): Boolean {
        return when (feature) {
            Feature.RestrictiveDataFiltering,
            Feature.Instrument,
            Feature.CrashReport,
            Feature.CrashShield,
            Feature.ThreadCheck,
            Feature.ErrorReport,
            Feature.AnrReport,
            Feature.AAM,
            Feature.CloudBridge,
            Feature.PrivacyProtection,
            Feature.SuggestedEvents,
            Feature.IntelligentIntegrity,
            Feature.ModelRequest,
            Feature.EventDeactivation,
            Feature.OnDeviceEventProcessing,
            Feature.OnDevicePostInstallEventProcessing,
            Feature.IapLogging,
            Feature.IapLoggingLib2,
            Feature.IapLoggingLib5To7,
            Feature.AndroidManualImplicitPurchaseDedupe,
            Feature.AndroidManualImplicitSubsDedupe,
            Feature.BannedParamFiltering,
            Feature.ProtectedMode,
            Feature.StdParamEnforcement,
            Feature.MACARuleMatching,
            Feature.BlocklistEvents,
            Feature.FilterRedactedEvents,
            Feature.FilterSensitiveParams,
            Feature.ChromeCustomTabsPrefetching,
            Feature.Monitoring,
            Feature.IgnoreAppSwitchToLoggedOut,
            Feature.BypassAppSwitch -> false

            else -> true
        }
    }

    /**
     * Feature enum Defines features in SDK
     *
     * Sample: AppEvents = 0x00010000, ^ ^ ^ ^ | | | | kit | | | feature | | sub-feature |
     * sub-sub-feature 1st byte: kit 2nd byte: feature 3rd byte: sub-feature 4th byte: sub-sub-feature
     */
    enum class Feature(private val code: Int) {
        Unknown(-1),

        // Features in CoreKit
        /** Essential of CoreKit */
        Core(0x00000000),
        AppEvents(0x00010000),
        CodelessEvents(0x00010100),
        CloudBridge(0x00010800),
        RestrictiveDataFiltering(0x00010200),
        AAM(0x00010300),
        PrivacyProtection(0x00010400),
        SuggestedEvents(0x00010401),
        IntelligentIntegrity(0x00010402),
        ModelRequest(0x00010403),
        ProtectedMode(0x00010404), /* filter out the params which are not supported legally */
        MACARuleMatching(0x00010405),
        BlocklistEvents(0x00010406), /* drop the events in the blocklist */
        FilterRedactedEvents(0x00010407), /* replace the event name via the redaction string */
        FilterSensitiveParams(0x00010408), /* filter out the sensitive params */
        StdParamEnforcement(0x01010409),
        BannedParamFiltering(0x0101040a),
        EventDeactivation(0x00010500),
        OnDeviceEventProcessing(0x00010600),
        OnDevicePostInstallEventProcessing(0x00010601),
        IapLogging(0x00010700),
        IapLoggingLib2(0x00010701),
        IapLoggingLib5To7(0x0010702),
        AndroidManualImplicitPurchaseDedupe(0x0010703),
        AndroidManualImplicitSubsDedupe(0x0010704),
        Instrument(0x00020000),
        CrashReport(0x00020100),
        CrashShield(0x00020101),
        ThreadCheck(0x00020102),
        ErrorReport(0x00020200),
        AnrReport(0x00020300),
        Monitoring(0x00030000),
        ServiceUpdateCompliance(0x00030100),
        Megatron(0x00040000),
        Elora(0x00050000),
        // Features in LoginKit
        /** Essential of LoginKit */
        Login(0x01000000),
        ChromeCustomTabsPrefetching(0x01010000),
        IgnoreAppSwitchToLoggedOut(0x01020000),
        BypassAppSwitch(0x01030000),
        // Features in ShareKit
        /** Essential of ShareKit */
        Share(0x02000000);

        override fun toString(): String =
            when (this) {
                Core -> "CoreKit"
                AppEvents -> "AppEvents"
                CodelessEvents -> "CodelessEvents"
                RestrictiveDataFiltering -> "RestrictiveDataFiltering"
                Instrument -> "Instrument"
                CrashReport -> "CrashReport"
                CrashShield -> "CrashShield"
                ThreadCheck -> "ThreadCheck"
                ErrorReport -> "ErrorReport"
                AnrReport -> "AnrReport"
                AAM -> "AAM"
                CloudBridge -> "AppEventsCloudbridge"
                PrivacyProtection -> "PrivacyProtection"
                SuggestedEvents -> "SuggestedEvents"
                IntelligentIntegrity -> "IntelligentIntegrity"
                StdParamEnforcement -> "StdParamEnforcement"
                ProtectedMode -> "ProtectedMode"
                BannedParamFiltering -> "BannedParamFiltering"
                MACARuleMatching -> "MACARuleMatching"
                BlocklistEvents -> "BlocklistEvents"
                FilterRedactedEvents -> "FilterRedactedEvents"
                FilterSensitiveParams -> "FilterSensitiveParams"
                ModelRequest -> "ModelRequest"
                EventDeactivation -> "EventDeactivation"
                OnDeviceEventProcessing -> "OnDeviceEventProcessing"
                OnDevicePostInstallEventProcessing -> "OnDevicePostInstallEventProcessing"
                IapLogging -> "IAPLogging"
                IapLoggingLib2 -> "IAPLoggingLib2"
                IapLoggingLib5To7 -> "IAPLoggingLib5To7"
                AndroidManualImplicitPurchaseDedupe -> "AndroidManualImplicitPurchaseDedupe"
                AndroidManualImplicitSubsDedupe -> "AndroidManualImplicitSubsDedupe"
                Monitoring -> "Monitoring"
                Megatron -> "Megatron"
                Elora -> "Elora"
                ServiceUpdateCompliance -> "ServiceUpdateCompliance"
                Login -> "LoginKit"
                ChromeCustomTabsPrefetching -> "ChromeCustomTabsPrefetching"
                IgnoreAppSwitchToLoggedOut -> "IgnoreAppSwitchToLoggedOut"
                BypassAppSwitch -> "BypassAppSwitch"
                Share -> "ShareKit"
                else -> "unknown"
            }

        fun toKey(): String {
            return "FBSDKFeature$this"
        }

        val parent: Feature
            get() =
                when {
                    this.code and 0xFF > 0 -> {
                        fromInt(this.code and -0x100)
                    }

                    this.code and 0xFF00 > 0 -> {
                        fromInt(this.code and -0x10000)
                    }

                    this.code and 0xFF0000 > 0 -> {
                        fromInt(this.code and -0x1000000)
                    }

                    else -> {
                        fromInt(0)
                    }
                }

        companion object {
            fun fromInt(code: Int): Feature {
                for (feature in values()) {
                    if (feature.code == code) {
                        return feature
                    }
                }
                return Unknown
            }
        }
    }

    /**
     * Callback for fetching feature status. Method [FeatureManager.checkFeature]} will call
     * GateKeeper manager to load the latest GKs first and then run the callback function.
     */
    fun interface Callback {
        /** The method that will be called when the feature status request completes. */
        fun onCompleted(enabled: Boolean)
    }
}
